File: crates/lae-core/src/service.rs
//! Task and taskspace orchestration — writes `state.db` and publishes bar updates.

use std::path::Path;

use chrono::Utc;

use crate::config::LaeConfig;
use crate::error::{LaeError, Result};
use crate::hyprland;
use crate::models::{ContextMode, SessionState, Task, TaskStatus};
use crate::registry::Registry;
use crate::state_notify::{self, StateChangeKind};
use crate::workspace_nav;
use crate::workspaces::{default_taskspace_workspace_names, task_taskspace_workspace_names};
use crate::xdg::{ensure_parent, lae_runtime_dir};

pub struct TaskService {
    registry: Registry,
    config: LaeConfig,
}

impl TaskService {
    pub fn with_defaults() -> Result<Self> {
        let config = crate::config::load_config()?;
        let registry = Registry::new(None, config.clone())?;
        Ok(Self { registry, config })
    }

    pub fn load_state(&self) -> Result<SessionState> {
        self.registry.load_state()
    }

    pub fn save_state(&self, state: &SessionState) -> Result<()> {
        self.commit_state(state, None)
    }

    /// Persist state and notify Waybar subscribers.
    fn commit_state(
        &self,
        state: &SessionState,
        change: Option<StateChangeKind>,
    ) -> Result<()> {
        self.registry.save_state(state)?;
        self.write_runtime_files(state)?;
        if let Some(kind) = change {
            state_notify::publish(kind);
        }
        Ok(())
    }

    /// Persist state after a workspace switch (no state-events publish).
    fn persist_workspace_switch(&self, state: &SessionState) -> Result<()> {
        self.registry.save_state(state)?;
        self.write_runtime_files(state)
    }

    fn write_runtime_files(&self, state: &SessionState) -> Result<()> {
        if let Ok(dir) = lae_runtime_dir() {
            ensure_parent(&dir.join("_"))?;
            let _ = std::fs::write(dir.join("context"), state.taskspace_label());
        }
        crate::workspace_slots::write_slot_cache(state);
        Ok(())
    }

    pub fn initialize(&self) -> Result<()> {
        let mut state = self.load_state()?;
        state.default_workspace_count = self.config.default_workspace_count;
        self.commit_state(&state, None)
    }

    /// Rename default numeric slots — run once after daemon start (background).
    pub fn provision_default_workspaces(&self) -> Result<()> {
        if hyprland::available() && self.config.hyprland_enabled {
            workspace_nav::setup_default_taskspace_workspaces(self.config.default_workspace_count);
        }
        Ok(())
    }

    /// Drop remembered last-workspace and per-monitor layout state.
    pub fn reset_navigation_layout(&self) -> Result<()> {
        let mut state = self.load_state()?;
        workspace_nav::clear_navigation_memory(&mut state);
        workspace_nav::clear_runtime_slot_cache();
        self.commit_state(&state, Some(StateChangeKind::Full))
    }

    pub fn context_default(&self) -> Result<()> {
        let mut state = self.load_state()?;
        workspace_nav::set_taskspace(&mut state, ContextMode::Default, None)
            .map_err(|e| crate::error::LaeError::Other(e))?;
        self.commit_state(&state, Some(StateChangeKind::Taskspace))
    }

    pub fn workspace_go(&self, relative: i32) -> Result<Option<String>> {
        let mut state = self.load_state()?;
        let name = workspace_nav::workspace_name_for_relative(&state, relative);
        if let Some(ref target) = name {
            hyprland::switch_workspace_for_navigation(target);
            workspace_nav::remember_workspace(&mut state, relative);
            self.persist_workspace_switch(&state)?;
        }
        Ok(name)
    }

    /// Hyprland-only workspace switch for keybind hot path (no state write).
    pub fn workspace_dispatch(&self, relative: i32) -> Result<Option<String>> {
        let name = crate::workspace_slots::read_slot_target(relative).or_else(|| {
            self.load_state()
                .ok()
                .and_then(|state| workspace_nav::workspace_name_for_relative(&state, relative))
        });
        if let Some(ref target) = name {
            hyprland::switch_workspace_for_navigation(target);
        }
        Ok(name)
    }

    pub fn remember_workspace_go(&self, relative: i32) -> Result<Option<String>> {
        let mut state = self.load_state()?;
        let name = workspace_nav::workspace_name_for_relative(&state, relative);
        if name.is_some() {
            workspace_nav::remember_workspace(&mut state, relative);
            self.persist_workspace_switch(&state)?;
        }
        Ok(name)
    }

    pub fn remember_workspace_goto(&self, name: &str) -> Result<Option<String>> {
        let mut state = self.load_state()?;
        let allowed = crate::workspaces::allowed_workspace_names(&state);
        if !allowed.iter().any(|n| n == name) {
            return Ok(None);
        }
        if let Some(idx) = allowed.iter().position(|n| n == name) {
            workspace_nav::remember_workspace(&mut state, (idx + 1) as i32);
            self.persist_workspace_switch(&state)?;
        }
        Ok(Some(name.to_string()))
    }

    pub fn workspace_next(&self) -> Result<Option<String>> {
        let state = self.load_state()?;
        let relative = workspace_nav::workspace_next_relative(&state);
        drop(state);
        if let Some(rel) = relative {
            return self.workspace_go(rel);
        }
        Ok(None)
    }

    pub fn workspace_prev(&self) -> Result<Option<String>> {
        let state = self.load_state()?;
        let relative = workspace_nav::workspace_prev_relative(&state);
        drop(state);
        if let Some(rel) = relative {
            return self.workspace_go(rel);
        }
        Ok(None)
    }

    pub fn workspace_goto(&self, name: &str) -> Result<Option<String>> {
        let mut state = self.load_state()?;
        let allowed = crate::workspaces::allowed_workspace_names(&state);
        if !allowed.iter().any(|n| n == name) {
            return Ok(None);
        }
        hyprland::switch_workspace_for_navigation(name);
        let allowed = crate::workspaces::allowed_workspace_names(&state);
        if let Some(idx) = allowed.iter().position(|n| n == name) {
            workspace_nav::remember_workspace(&mut state, (idx + 1) as i32);
            self.persist_workspace_switch(&state)?;
        }
        Ok(Some(name.to_string()))
    }

    pub fn create_task(
        &self,
        name: &str,
        switch: bool,
        repo: crate::task_repo::TaskRepoSource,
        cwd: Option<&Path>,
    ) -> Result<Task> {
        let mut state = self.load_state()?;
        let active_count = state
            .tasks
            .values()
            .filter(|t| t.status != TaskStatus::Archived)
            .count();
        if active_count >= self.config.max_tasks as usize {
            return Err(LaeError::Other(format!(
                "Maximum task limit ({}) reached",
                self.config.max_tasks
            )));
        }

        let task_id = self.registry.unique_task_id(&state, name);
        let task_home = self.config.tasks_base_dir.join(&task_id);

        let (repo_path, create_repo_dir) = repo.resolve(&task_home, cwd)?;
        let repo_url = repo.resolve_url();

        let agent_dir = task_home.join(".lae");
        std::fs::create_dir_all(&agent_dir).map_err(|source| LaeError::Write {
            path: agent_dir.clone(),
            source,
        })?;
        let notes_path = agent_dir.join("agent-notes.md");
        if !notes_path.is_file() {
            std::fs::write(
                &notes_path,
                format!("# {name}\n\nTask notes for agent and human.\n"),
            )
            .map_err(|source| LaeError::Write {
                path: notes_path.clone(),
                source,
            })?;
        }
        if create_repo_dir {
            std::fs::create_dir_all(&repo_path).map_err(|source| LaeError::Write {
                path: repo_path.clone(),
                source,
            })?;
        }

        let now = Utc::now();
        let mut task = Task {
            id: task_id.clone(),
            name: name.to_string(),
            status: TaskStatus::Active,
            repo_url,
            repo_path,
            branch: None,
            container_name: format!("{}-{task_id}", self.config.container_prefix),
            workspace_count: self.config.default_workspace_count,
            browser_profile: None,
            created_at: now,
            last_active_at: now,
            agent_notes_path: Some(notes_path),
            ports: vec![],
        };
        self.registry.touch_task(&mut task);
        state.tasks.insert(task_id.clone(), task.clone());

        if hyprland::available() && self.config.hyprland_enabled && !switch {
            workspace_nav::setup_task_workspaces(&task_id, self.config.default_workspace_count);
        }

        if switch {
            self.commit_state(&state, Some(StateChangeKind::Full))?;
            self.switch_task(&task_id)
        } else {
            self.commit_state(&state, Some(StateChangeKind::Full))?;
            Ok(task)
        }
    }

    pub fn archive_task(&self, task_id: &str) -> Result<()> {
        let mut state = self.load_state()?;
        let task = state
            .tasks
            .get(task_id)
            .cloned()
            .ok_or_else(|| LaeError::Other(format!("Unknown task: {task_id}")))?;
        if task.status == TaskStatus::Archived {
            return Err(LaeError::Other(format!("Task is already archived: {task_id}")));
        }

        let was_current = crate::task_cleanup::is_active_task_context(&state, &task);
        if was_current {
            workspace_nav::set_taskspace(&mut state, ContextMode::Default, None)
                .map_err(LaeError::Other)?;
            self.commit_state(&state, Some(StateChangeKind::Taskspace))?;
        }

        let _closed = crate::task_cleanup::close_task_windows(&task)?;
        if let Err(err) = crate::task_cleanup::stop_task_container(&task) {
            eprintln!(
                "lae: archive task {}: stop container {}: {err}",
                task.id, task.container_name
            );
        }
        crate::task_cleanup::purge_task_windows(&mut state, task_id);

        if let Some(entry) = state.tasks.get_mut(task_id) {
            entry.status = TaskStatus::Archived;
        }

        self.commit_state(&state, Some(StateChangeKind::Full))
    }

    pub fn delete_task(&self, task_id: &str) -> Result<()> {
        let mut state = self.load_state()?;
        let task = state
            .tasks
            .get(task_id)
            .cloned()
            .ok_or_else(|| LaeError::Other(format!("Unknown task: {task_id}")))?;

        if crate::task_cleanup::is_active_task_context(&state, &task) {
            workspace_nav::set_taskspace(&mut state, ContextMode::Default, None)
                .map_err(LaeError::Other)?;
            self.commit_state(&state, Some(StateChangeKind::Taskspace))?;
        }

        let _closed = crate::task_cleanup::close_task_windows(&task)?;
        if let Err(err) = crate::task_cleanup::stop_task_container(&task) {
            eprintln!(
                "lae: delete task {}: stop container {}: {err}",
                task.id, task.container_name
            );
        }
        if let Err(err) = crate::task_cleanup::remove_task_container(&task) {
            eprintln!(
                "lae: delete task {}: remove container {}: {err}",
                task.id, task.container_name
            );
        }
        if let Err(err) = crate::task_cleanup::remove_task_data_dir(&self.config, &task) {
            eprintln!("lae: delete task {}: remove data dir: {err}", task.id);
        }
        crate::task_cleanup::purge_task_windows(&mut state, task_id);
        crate::task_cleanup::purge_task_session_keys(&mut state, task_id);
        state.tasks.remove(task_id);

        self.commit_state(&state, Some(StateChangeKind::Full))
    }

    pub fn preview_task_teardown(&self, task_id: &str) -> Result<crate::task_cleanup::TaskTeardownPreview> {
        let state = self.load_state()?;
        let task = state
            .tasks
            .get(task_id)
            .cloned()
            .ok_or_else(|| LaeError::Other(format!("Unknown task: {task_id}")))?;
        crate::task_cleanup::preview_teardown(&self.config, &task)
    }

    pub fn list_archived_tasks(&self) -> Result<Vec<Task>> {
        let state = self.load_state()?;
        Ok(state
            .tasks
            .values()
            .filter(|t| t.status == TaskStatus::Archived)
            .cloned()
            .collect())
    }

    pub fn switch_task(&self, task_id: &str) -> Result<Task> {
        let mut state = self.load_state()?;
        let task = state
            .tasks
            .get(task_id)
            .cloned()
            .ok_or_else(|| crate::error::LaeError::Other(format!("Unknown task: {task_id}")))?;
        if task.status == TaskStatus::Archived {
            return Err(crate::error::LaeError::Other(format!(
                "Task is archived: {task_id}"
            )));
        }
        let mut task = task;
        task.status = TaskStatus::Active;
        task.last_active_at = Utc::now();
        state.tasks.insert(task.id.clone(), task.clone());

        workspace_nav::set_taskspace(&mut state, ContextMode::Task, Some(task_id))
            .map_err(|e| crate::error::LaeError::Other(e))?;
        state
            .last_workspace
            .entry(format!("task:{task_id}"))
            .or_insert(1);

        self.commit_state(&state, Some(StateChangeKind::Taskspace))?;
        Ok(task)
    }

    pub fn resolve_task(&self, name_or_id: &str) -> Result<Task> {
        let state = self.load_state()?;
        self.registry
            .get_task(&state, name_or_id)
            .cloned()
            .ok_or_else(|| crate::error::LaeError::Other(format!("Unknown task: {name_or_id}")))
    }

    pub fn list_active_tasks(&self) -> Result<Vec<Task>> {
        let state = self.load_state()?;
        Ok(state
            .tasks
            .values()
            .filter(|t| t.status != TaskStatus::Archived)
            .cloned()
            .collect())
    }

    pub fn tasks_for_menu(&self) -> Result<Vec<MenuTask>> {
        let state = self.load_state()?;
        let mut items = Vec::new();

        items.push(MenuTask {
            id: "default".into(),
            name: "default".into(),
            kind: "default".into(),
            workspaces: default_taskspace_workspace_names(state.default_workspace_count),
            current: state.context_mode == ContextMode::Default,
            status: "system".into(),
        });

        for task in state.tasks.values() {
            if task.status == TaskStatus::Archived {
                continue;
            }
            items.push(MenuTask {
                id: task.id.clone(),
                name: task.name.clone(),
                kind: "task".into(),
                workspaces: task_taskspace_workspace_names(&state, &task.id),
                current: state.context_mode == ContextMode::Task
                    && state.current_task_id.as_deref() == Some(task.id.as_str()),
                status: task.status.as_str().into(),
            });
        }
        Ok(items)
    }

    pub fn taskspace_label(&self) -> Result<String> {
        Ok(self.load_state()?.taskspace_label())
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct MenuTask {
    pub id: String,
    pub name: String,
    pub kind: String,
    pub workspaces: Vec<String>,
    pub current: bool,
    pub status: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::LaeConfig;
    use tempfile::tempdir;

    fn test_service(dir: &std::path::Path) -> TaskService {
        let mut config = LaeConfig::default();
        config.tasks_base_dir = dir.join("tasks");
        let db = dir.join("state.db");
        let registry = Registry::new(Some(db), config.clone()).unwrap();
        TaskService { registry, config }
    }

    #[test]
    fn create_task_registers_workspaces_without_switch() {
        let dir = tempdir().unwrap();
        let svc = test_service(dir.path());
        let task = svc
            .create_task(
                "Auth Fix",
                false,
                crate::task_repo::TaskRepoSource::Scratch,
                None,
            )
            .unwrap();
        assert!(task.id.starts_with('t'));
        assert_eq!(task.name, "Auth Fix");
        assert_eq!(task.workspace_count, 10);
        assert_eq!(task.workspace_names().len(), 10);
        assert_eq!(task.workspace_names()[0], format!("{}-1", task.id));
        assert_eq!(task.workspace_names()[9], format!("{}-10", task.id));
        assert!(task.agent_notes_path.as_ref().unwrap().is_file());
        assert!(task.repo_path.is_dir());

        let state = svc.load_state().unwrap();
        assert!(state.tasks.contains_key(&task.id));
        assert_eq!(state.context_mode, ContextMode::Default);
    }

    #[test]
    fn create_task_with_switch_enters_taskspace() {
        let dir = tempdir().unwrap();
        let svc = test_service(dir.path());
        let task = svc
            .create_task(
                "billing",
                true,
                crate::task_repo::TaskRepoSource::Scratch,
                None,
            )
            .unwrap();
        let state = svc.load_state().unwrap();
        assert_eq!(state.context_mode, ContextMode::Task);
        assert_eq!(state.current_task_id.as_deref(), Some(task.id.as_str()));
    }

    #[test]
    fn archive_task_leaves_default_taskspace() {
        let dir = tempdir().unwrap();
        let svc = test_service(dir.path());
        let task = svc
            .create_task(
                "temp",
                true,
                crate::task_repo::TaskRepoSource::Scratch,
                None,
            )
            .unwrap();
        svc.archive_task(&task.id).unwrap();
        let state = svc.load_state().unwrap();
        assert_eq!(state.tasks.get(&task.id).unwrap().status, TaskStatus::Archived);
        assert_eq!(state.context_mode, ContextMode::Default);
        assert!(state.current_task_id.is_none());
        assert!(svc.tasks_for_menu().unwrap().iter().all(|t| t.id != task.id));
        assert!(dir.path().join("tasks").join(&task.id).is_dir());
    }

    #[test]
    fn delete_task_removes_record_and_data() {
        let dir = tempdir().unwrap();
        let svc = test_service(dir.path());
        let task = svc
            .create_task(
                "gone",
                false,
                crate::task_repo::TaskRepoSource::Scratch,
                None,
            )
            .unwrap();
        let task_home = dir.path().join("tasks").join(&task.id);
        assert!(task_home.is_dir());

        svc.delete_task(&task.id).unwrap();
        let state = svc.load_state().unwrap();
        assert!(!state.tasks.contains_key(&task.id));
        assert!(!task_home.exists());
    }

    #[test]
    fn delete_task_while_current_leaves_default_taskspace() {
        let dir = tempdir().unwrap();
        let svc = test_service(dir.path());
        let task = svc
            .create_task(
                "current",
                true,
                crate::task_repo::TaskRepoSource::Scratch,
                None,
            )
            .unwrap();
        svc.delete_task(&task.id).unwrap();
        let state = svc.load_state().unwrap();
        assert!(!state.tasks.contains_key(&task.id));
        assert_eq!(state.context_mode, ContextMode::Default);
        assert!(state.current_task_id.is_none());
    }

    #[test]
    fn create_task_with_path_uses_checkout_not_scratch() {
        let dir = tempdir().unwrap();
        let checkout = dir.path().join("checkout");
        std::fs::create_dir_all(checkout.join(".git")).unwrap();
        let svc = test_service(dir.path());
        let task = svc
            .create_task(
                "My Feature",
                false,
                crate::task_repo::TaskRepoSource::Path(checkout.clone()),
                None,
            )
            .unwrap();
        assert_eq!(task.name, "My Feature");
        assert!(task.id.starts_with('t'));
        assert_eq!(task.repo_path, checkout);
        let scratch = dir.path().join("tasks").join(&task.id).join("repo");
        assert!(!scratch.exists() || !scratch.is_dir());
    }
}

File: src/lae/daemon/service.py
"""Task and context orchestration."""

from __future__ import annotations

import os
import subprocess
from pathlib import Path

from lae.util import subprocess as sp

from lae.core.config import LaeConfig, load_config
from lae.core.models import ContextMode, SessionState, Task, TaskStatus
from lae.core.registry import Registry
from lae.core.workspaces import default_taskspace_workspace_names
from lae.daemon import workspace_nav
from lae.daemon.waybar_export import write_waybar_file
from lae.integrations import distrobox, git, hyprland
from lae.util import xdg


class TaskService:
    def __init__(
        self,
        registry: Registry | None = None,
        config: LaeConfig | None = None,
    ):
        self.config = config or load_config()
        self.registry = registry or Registry(config=self.config)
        self._state: SessionState | None = None

    def get_state(self) -> SessionState:
        # Always reload — Rust CLI and Waybar CFFI write state.db directly.
        self._state = self.registry.load_state()
        return self._state

    def save_state(self, state: SessionState | None = None) -> None:
        state = state or self.get_state()
        self.registry.save_state(state)
        self._write_runtime_files(state)

    def _write_runtime_files(self, state: SessionState) -> None:
        self._write_context_file(state)
        write_waybar_file(state)

    def _write_context_file(self, state: SessionState) -> None:
        try:
            runtime = xdg.lae_runtime_dir()
            runtime.mkdir(parents=True, exist_ok=True)
            xdg.lae_context_file().write_text(state.taskspace_label())
        except RuntimeError:
            pass

    def initialize(self) -> None:
        state = self.get_state()
        state.default_workspace_count = self.config.default_workspace_count
        if hyprland.available() and self.config.hyprland_enabled:
            workspace_nav.setup_default_taskspace_workspaces(
                self.config.default_workspace_count
            )
            for task in state.tasks.values():
                if task.status != TaskStatus.archived:
                    workspace_nav.setup_task_workspaces(
                        task, slot_count=state.default_workspace_count
                    )
        self.save_state(state)

    def create_task(
        self,
        name: str,
        *,
        repo_url: str | None = None,
        branch: str | None = None,
        switch: bool = True,
    ) -> Task:
        state = self.get_state()
        task_id = self.registry.unique_task_id(state, name)

        task_home = self.config.tasks_base_dir / task_id
        repo_path = task_home / "repo"
        agent_dir = task_home / ".lae"
        agent_dir.mkdir(parents=True, exist_ok=True)
        notes_path = agent_dir / "agent-notes.md"
        if not notes_path.exists():
            notes_path.write_text(f"# {name}\n\nTask notes for agent and human.\n")

        if repo_url:
            git.clone_repo(repo_url, repo_path, branch)
        else:
            git.init_repo(repo_path)

        container_name = f"{self.config.container_prefix}-{task_id}"
        if distrobox.available():
            if not distrobox.container_exists(container_name):
                distrobox.create_container(
                    container_name,
                    task_home,
                    self.config.distrobox_image,
                )

        task = Task(
            id=task_id,
            name=name,
            status=TaskStatus.active,
            repo_url=repo_url,
            repo_path=repo_path,
            branch=branch,
            container_name=container_name,
            workspace_count=self.config.default_workspace_count,
            agent_notes_path=notes_path,
        )
        state.tasks[task_id] = task
        self.registry.touch_task(task)

        if hyprland.available() and self.config.hyprland_enabled:
            workspace_nav.setup_task_workspaces(
                task, slot_count=state.default_workspace_count
            )

        if switch:
            self.switch_task(task_id)
        else:
            self.save_state(state)
        return task

    def switch_task(self, task_id: str) -> Task:
        state = self.get_state()
        task = state.tasks.get(task_id)
        if task is None:
            raise ValueError(f"Unknown task: {task_id}")
        if task.status == TaskStatus.archived:
            raise ValueError(f"Task is archived: {task_id}")
        task.status = TaskStatus.active
        self.registry.touch_task(task)
        workspace_nav.set_taskspace(state, ContextMode.task, task_id)
        state.last_workspace[f"task:{task_id}"] = state.last_workspace.get(
            f"task:{task_id}", 1
        )
        if hyprland.available():
            workspace_nav.setup_task_workspaces(
                task, slot_count=state.default_workspace_count
            )
            hyprland.switch_workspace(task.main_workspace())
        self.save_state(state)
        return task

    def context_default(self) -> None:
        state = self.get_state()
        workspace_nav.set_taskspace(state, ContextMode.default)
        self.save_state(state)

    def archive_task(self, task_id: str) -> None:
        from lae.core import task_cleanup

        state = self.get_state()
        task = state.tasks.get(task_id)
        if task is None:
            raise ValueError(f"Unknown task: {task_id}")
        if task.status == TaskStatus.archived:
            raise ValueError(f"Task is already archived: {task_id}")

        if state.current_task_id == task_id:
            workspace_nav.set_taskspace(state, ContextMode.default)
            state.current_task_id = None

        task_cleanup.close_task_windows(task)
        task_cleanup.stop_task_container(task)
        task_cleanup.purge_task_windows(state, task_id)
        task.status = TaskStatus.archived
        self.save_state(state)

    def delete_task(self, task_id: str) -> None:
        from lae.core import task_cleanup

        state = self.get_state()
        task = state.tasks.get(task_id)
        if task is None:
            raise ValueError(f"Unknown task: {task_id}")

        if state.current_task_id == task_id:
            workspace_nav.set_taskspace(state, ContextMode.default)
            state.current_task_id = None

        task_cleanup.close_task_windows(task)
        task_cleanup.remove_task_container(task)
        task_cleanup.remove_task_data_dir(self.config.tasks_base_dir, task)
        task_cleanup.purge_task_windows(state, task_id)
        task_cleanup.purge_task_session_keys(state, task_id)
        del state.tasks[task_id]
        self.save_state(state)

    def open_terminal(
        self,
        task_id: str | None = None,
        *,
        host: bool = False,
    ) -> None:
        state = self.get_state()
        if host:
            self._launch_host_terminal(None)
            return

        tid = task_id or state.current_task_id
        if not tid:
            raise RuntimeError(
                "No task context. Run `lae task switch <name>` or pass a task name."
            )
        task = state.tasks.get(tid)
        if task is None:
            raise ValueError(f"Unknown task: {tid}")

        inner = f"cd ~/repo && exec ${{'SHELL:-/bin/bash'}}"
        enter_argv = distrobox.enter_command(task.container_name, inner)
        title = f"[{task.id}] terminal"
        self._launch_terminal(task, enter_argv, title)

    def _launch_terminal(
        self, task: Task | None, command: list[str], title: str
    ) -> None:
        term = self.config.terminal_command
        if not sp.which(term):
            raise RuntimeError(f"Terminal emulator not found: {term}")

        argv = [term]
        if self.config.terminal_title_flag:
            argv.extend([self.config.terminal_title_flag, title])
        if task:
            argv.extend(["--working-directory", str(task.repo_path)])
        argv.extend(["--", *command])
        subprocess.Popen(argv, env=os.environ.copy())

    def _launch_host_terminal(self, cwd: Path | None) -> None:
        term = self.config.terminal_command
        if not sp.which(term):
            raise RuntimeError(f"Terminal emulator not found: {term}")
        argv = [term]
        if cwd:
            argv.extend(["--working-directory", str(cwd)])
        subprocess.Popen(argv, env=os.environ.copy())

    def resolve_task(self, name_or_id: str) -> Task:
        state = self.get_state()
        task = self.registry.get_task(state, name_or_id)
        if task is None:
            raise ValueError(f"Unknown task: {name_or_id}")
        return task

    def list_active_tasks(self) -> list[Task]:
        state = self.get_state()
        return [t for t in state.tasks.values() if t.status != TaskStatus.archived]

    def tasks_for_menu(self) -> list[dict]:
        state = self.get_state()
        items: list[dict] = []

        in_default = state.context_mode == ContextMode.default
        items.append(
            {
                "id": "default",
                "name": "default",
                "kind": "default",
                "workspaces": default_taskspace_workspace_names(
                    state.default_workspace_count
                ),
                "current": in_default,
                "status": "system",
            }
        )

        for task in self.list_active_tasks():
            items.append(
                {
                    "id": task.id,
                    "name": task.name,
                    "kind": "task",
                    "workspaces": task.workspace_names(),
                    "current": (
                        state.context_mode == ContextMode.task
                        and state.current_task_id == task.id
                    ),
                    "status": task.status.value,
                }
            )
        return items

    def status_summary(self) -> dict:
        state = self.get_state()
        active_ws = hyprland.get_active_workspace() if hyprland.available() else None
        current_task = (
            state.tasks.get(state.current_task_id) if state.current_task_id else None
        )
        allowed = workspace_nav.allowed_workspaces(state)
        branch = git.current_branch(current_task.repo_path) if current_task else None
        container_running = (
            distrobox.is_running(current_task.container_name) if current_task else False
        )
        return {
            "taskspace": state.taskspace_label(),
            "context": state.taskspace_label(),
            "allowed_workspaces": allowed,
            "active_workspace": active_ws.name if active_ws else None,
            "active_workspace_id": active_ws.id if active_ws else None,
            "current_task": current_task.model_dump(mode="json") if current_task else None,
            "branch": branch,
            "container_running": container_running,
            "tasks": [t.model_dump(mode="json") for t in self.list_active_tasks()],
            "windows": [
                w.model_dump(mode="json", by_alias=True) for w in state.windows.values()
            ],
        }
